src/wasm/mod.rs

```diff
@@ -0,0 +1,233 @@
+                    _ => Err(wasmi::Error::Instantiation("unresolved func import".to_owned()))
+                }
+            }
+
+            fn signature(self) -> wasmi::Signature {
+                match self {
+                    $(
+                        HostFunc::$variant => wasmi::Signature::new(
+                            &[$(<$t as WasmPrimitive>::TYPE),*][..],
+                            option_helper!(Some $(<$rt as WasmPrimitive>::TYPE)?),
+                        )
+                    ),*
+                }
+            }
+
+            fn func_ref(self) -> wasmi::FuncRef {
+                wasmi::FuncInstance::alloc_host(self.signature(), self as usize)
+            }
+
+            fn module_name(self) -> &'static str {
+                match self {
+                    $(HostFunc::$variant => $module),*
+                }
+            }
+
+            fn field_name(self) -> &'static str {
+                match self {
+                    $(HostFunc::$variant => $name),*
+                }
+            }
+        }
+
+        impl TryFrom<usize> for HostFunc {
+            type Error = wasmi::Trap;
+            fn try_from(x: usize) -> Result<Self, Self::Error> {
+                $(
+                    if x == (HostFunc::$variant as usize) {
+                        return Ok(HostFunc::$variant);
+                    }
+                )*
+                Err(wasmi::TrapKind::UnexpectedSignature.into())
+            }
+        }
+
+        impl fmt::Display for HostFunc {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "{}::{}", self.module_name(), self.field_name())
+            }
+        }
+
+        impl wasmi::Externals for Host {
+            fn invoke_index(
+                &mut self,
+                index: usize,
+                args: wasmi::RuntimeArgs,
+            ) -> Result<Option<wasmi::RuntimeValue>, wasmi::Trap> {
+                let span = tracing::trace_span!("invoke_index", index, ?args);
+                let _enter = span.enter();
+
+                match HostFunc::try_from(index)? {
+                    $(
+                        HostFunc::$variant => match args.as_ref() {
+                            [$($p),*] => {
+                                let _result = $method(
+                                    self,
+                                    $(<$t as WasmPrimitive>::from_wasm_value(*$p)?),*
+                                )?;
+                                Ok(option_helper!(
+                                    Some $(<$rt as WasmPrimitive>::into_wasm_value(_result))?
+                                ))
+                            }
+                            _ => Err(wasmi::TrapKind::UnexpectedSignature.into()),
+                        }
+                    ),*
+                }
+            }
+        }
+    }
+}
+
+host_funcs! {
+    fn "wasi_unstable"::"fd_write"(fd: u32, iovs: u32, iovs_len: u32, nwritten: u32) -> u16
```

Comment:
TIOLI: seems like these could be idents rather than quoted?

Reply:
*shrug* probably.
I made them strings because I think _theoretically_ wasm supports spaces in symbols. I'm not sure we'll ever declare a symbol which has spaces in it, though, so it's probably not reasonable.